File: src/main/java/me/newyith/fortress/main/FortressPlugin.java
package me.newyith.fortress.main;

public class FortressPlugin {
//	public static final boolean releaseBuild = false; //TODO: change this to true for release builds
//
//	private static FortressPlugin instance;
//	private static SaveLoadManager saveLoadManager;
//	private static SandboxSaveLoadManager sandboxSaveLoadManager;
//
//	public static int config_glowstoneDustBurnTimeMs = 1000 * 60 * 60;
//	public static int config_stuckDelayMs = 30 * 1000;
//	public static int config_stuckCancelDistance = 4;
//	public static int config_generationRangeLimit = 128;
//	public static int config_generationBlockLimit = 40000; //roughly 125 empty 8x8x8 rooms (6x6x6 air inside)
//
//	private void loadConfig() {
//		FileConfiguration config = getConfig();
//		if (releaseBuild) {
//			config_glowstoneDustBurnTimeMs = getConfigInt(config, "glowstoneDustBurnTimeMs", config_glowstoneDustBurnTimeMs);
//			config_stuckDelayMs = getConfigInt(config, "stuckDelayMs", config_stuckDelayMs);
//			config_stuckCancelDistance = getConfigInt(config, "stuckCancelDistance", config_stuckCancelDistance);
//			config_generationRangeLimit = getConfigInt(config, "generationRangeLimit", config_generationRangeLimit);
//			config_generationBlockLimit = getConfigInt(config, "generationBlockLimit", config_generationBlockLimit);
//		}
//		saveConfig();
//	}
//	private int getConfigInt(FileConfiguration config, String key, int defaultValue) {
//		if (!config.isInt(key)) {
//			config.set(key, defaultValue);
//		}
//		return config.getInt(key);
//	}
//
//	public void onEnable() {
//		instance = this;
//
//		loadConfig();
//
//		Log.sendConsole("%%%%%%%%%%%%%%%%%%%%%%%%%%%%", ChatColor.RED);
//		Log.sendConsole(">>    Fortress Plugin     <<", ChatColor.GOLD);
//		Log.sendConsole("         >> ON <<           ", ChatColor.GREEN);
//		Log.sendConsole("%%%%%%%%%%%%%%%%%%%%%%%%%%%%", ChatColor.RED);
//
//		saveLoadManager = new SaveLoadManager(this);
//		saveLoadManager.load();
//
//		if (!releaseBuild) {
//			sandboxSaveLoadManager = new SandboxSaveLoadManager(this);
////			sandboxSaveLoadManager.load();
//		}
//
//		EventListener.onEnable(this);
//		TickTimer.onEnable(this);
//		ManualCraftManager.onEnable(this);
//		PearlGlitchFix.onEnable(this);
//	}
//
//	public void onDisable() {
//		saveLoadManager.save();
//		if (!releaseBuild) {
////			sandboxSaveLoadManager.save();
//		}
//
//		Log.sendConsole("%%%%%%%%%%%%%%%%%%%%%%%%%%%%", ChatColor.RED);
//		Log.sendConsole(">>    Fortress Plugin     <<", ChatColor.GOLD);
//		Log.sendConsole("         >> OFF <<          ", ChatColor.RED);
//		Log.sendConsole("%%%%%%%%%%%%%%%%%%%%%%%%%%%%", ChatColor.RED);
//	}
//
//	public boolean onCommand(CommandSender sender, Command cmd, String label, String[] args) {
//		String commandName = cmd.getName();
//		boolean commandHandled = false;
//
//		// /stuck
//		if (commandName.equalsIgnoreCase("stuck") && sender instanceof Player) {
//			Player player = (Player)sender;
//			Commands.onStuckCommand(player);
//			commandHandled = true;
//		}
//
//		return commandHandled;
//	}
}

File: src/main/java/me/newyith/fortress/main/FortressPlugin.kt
package me.newyith.fortress.main

import me.newyith.fortress.config.ConfigData
import me.newyith.fortress.config.ConfigManager
import me.newyith.fortress.persist.SaveLoad
import me.newyith.util.Log
import org.bukkit.ChatColor
import org.bukkit.World
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.plugin.java.JavaPlugin

/**
 * Responsible for saving/loading FortressPluginForWorlds.
 * Also provides centralized access to some globals such as isRelease.
 * Must set all its fields to null onDisable() to avoid memory leak when using /reload command.
 */
object FortressPlugin {
	val releaseBuild get() = false //TODO: change this to true for release builds
	var pluginForWorldsReal: FortressPluginForWorlds? = null
	private var saveLoadReal: SaveLoad? = null
	private var config: ConfigData? = null
	private var plugin: JavaPlugin? = null

	// save & load //

	private val pluginForWorlds get() = getOrLoadOrCreatePluginForWorlds()
	val saveLoad get() = getOrCreateSaveLoad()

	fun getOrLoadOrCreatePluginForWorlds(): FortressPluginForWorlds {
		val path by lazy { SaveLoad.getSavePathOfFortressPluginForWorlds() }
		val pluginForWorlds = pluginForWorldsReal
			?: saveLoad.load(path)
			?: FortressPluginForWorlds()
		pluginForWorldsReal = pluginForWorlds
		return pluginForWorlds
	}

	fun getOrCreateSaveLoad(): SaveLoad {
		val saveLoad = saveLoadReal ?: SaveLoad()
		saveLoadReal = saveLoad
		return saveLoad
	}

	// getters //

	fun getConfig(): ConfigData {
		return config ?: ConfigManager.getDefaults()
	}

	fun getPlugin(): JavaPlugin? {
		return plugin
	}

	fun forWorld(world: World): FortressPluginForWorld {
		return pluginForWorlds.forWorld(world)
	}

	// handlers //

	fun onEnable(javaPlugin: JavaPlugin) {
		Log.sendConsole("%%%%%%%%%%%%%%%%%%%%%%%%%%%%", ChatColor.RED)
		Log.sendConsole(">>    Fortress Plugin     <<", ChatColor.GOLD)

		FortressPlugin.config = ConfigManager.loadOrSave(javaPlugin)
		FortressPlugin.plugin = javaPlugin

		//load or create pluginForWorlds
		getOrLoadOrCreatePluginForWorlds() //this line is not strictly necessary
		pluginForWorlds.onEnable(javaPlugin)

		Log.sendConsole("         >> ON <<           ", ChatColor.GREEN)
		Log.sendConsole("%%%%%%%%%%%%%%%%%%%%%%%%%%%%", ChatColor.RED)
	}

	fun onDisable() {
		Log.sendConsole("%%%%%%%%%%%%%%%%%%%%%%%%%%%%", ChatColor.RED)
		Log.sendConsole(">>    Fortress Plugin     <<", ChatColor.GOLD)

		//save pluginForWorlds
		val path = SaveLoad.getSavePathOfFortressPluginForWorlds()
		saveLoad.save(pluginForWorlds, path)
		pluginForWorlds.onDisable()

		FortressPlugin.pluginForWorldsReal = null
		FortressPlugin.saveLoadReal = null
		FortressPlugin.config = null
		FortressPlugin.plugin = null

		Log.sendConsole("         >> OFF <<          ", ChatColor.RED)
		Log.sendConsole("%%%%%%%%%%%%%%%%%%%%%%%%%%%%", ChatColor.RED)
	}

	fun onCommand(sender: CommandSender, cmd: Command, label: String, args: Array<String>): Boolean {
		return pluginForWorlds.onCommand(sender, cmd, label, args)
	}

	fun onTick() {
		pluginForWorlds.onTick()
	}
}